server/src/main/java/org/elasticsearch/plugins/ActionPlugin.java

```diff
@@ -48,29 +47,40 @@ public interface ActionPlugin {
     /**
      * Actions added by this plugin.
      */
-    default List<ActionHandler<? extends ActionRequest, ? extends ActionResponse>> getActions() {
+    default Collection<ActionHandler<? extends ActionRequest, ? extends ActionResponse>> getActions() {
         return Collections.emptyList();
     }
 
     /**
      * ActionType filters added by this plugin.
      */
-    default List<ActionFilter> getActionFilters() {
+    default Collection<ActionFilter> getActionFilters() {
         return Collections.emptyList();
     }
 
+    /**
+     * Encapsulates parameters to the {@link #getRestHandlers} method.
+     */
+    interface RestHandlerParameters {
```

Comment:
Briefly following up on the discussion in https://github.com/elastic/elasticsearch/pull/103277.

What I like about the change beyond cosmetics is that it makes code analysis slightly easier. 

But, on the other hand, I totally agree with the concern raised that such param objects make it a lot easier to silently and sneakishly grow the scope of access by extending this with minimal changes necessary. Not sure if an appropriate comment here would be discouraging enough to prevent that, but likely (and usually) not.

Reply:
I think it actually makes adding a parameter more difficult - it's a single refactor in intellij to add a new parameter to a method. Adding a new parameter to this class is more involved and can't be done programmatically, it has to be more deliberate.

Reply:
In terms of implementation effort, yes, I agree. But I think what's more relevant in that context is the perceived size / complexity of a change in terms of number of files / lines change on the respective PR.

Reply:
The perceived complexity of adding new rest handler parameters is exactly the problem: if a PR is small, it is often overlooked as trivial and therefore not given much thought. Yet a change to this rest handler signature reflects the impact: we should not be freely adding new parameters here.

Reply:
FWIW I don't think this argument is particularly robust. It's easy enough to add a parameter here as needed whether we have a parameter object or not; we don't have any docs today saying to avoid this, or even what to consider before doing so, and I doubt such a change would get a core/infra review request if it was part of some other team's dev work.

In support of adding a parameter object here: if we ever do add a parameter, then the impact on backports can be quite significant. It basically breaks auto-backports of changes that do something near to this signature, and I think that introduces some risk of introducing more bugs.

Also I note that we made such a change recently in #101684, and then had to follow up with #105022 and then again with #105101. I don't think these obstructions really made us think again about whether this was the right thing to do, they're just toil, and if we had had a parameter object in this area then we'd have avoided all that.

Reply:
Likewise https://github.com/elastic/elasticsearch/pull/104851/commits/33fa0cdb2448db09ca9791f097a70a0eb8f06f6d - I mean it's only a couple of minutes to fix this and re-run precommit but still it feels like unnecessary friction in the process.